tools/clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -5097,14 +5097,65 @@ static void ProcessDeclAttribute(Sema &S, Scope *scope, Decl *D,
   }
 }
 
+void Sema::ValidateShaderAttribtues(Scope *S, Decl *D, const AttributeList *A) {
+  bool hasComputeAttr = false;
+  bool hasNodeAttr = false;
+  int attrNum = 0;
+
+  for (const AttributeList *l = A; l; l = l->getNext()) {
+    if (l->getKind() == AttributeList::AT_HLSLShader) {
+      Expr *E = l->getArgAsExpr(0);
+      if (E->isTypeDependent() || E->isValueDependent() ||
+          E->getStmtClass() != Stmt::StringLiteralClass) {
+        break;
+      }
+
+      StringLiteral *sl = cast<StringLiteral>(E);
+      std::string str = sl->getString();
```

Comment:
Why use a `std::string` here? `st->getString()` returns a `StringRef` which should be sufficient here and is more efficient.